src/GravityVisitor.h

```diff
@@ -14,12 +14,74 @@ class GravityVisitor {
   static constexpr int  nMinParticleNode = 6;
 
   void addGravity(const SpatialNode<CentroidData>& source, SpatialNode<CentroidData>& target) {
-    for (int i = 0; i < target.n_particles; i++) {
-      Vector3D<Real> diff = source.data.centroid - target.particles()[i].position;
-      Real rsq = diff.lengthSquared();
-      if (rsq != 0) {
-        Vector3D<Real> accel = diff * (source.data.sum_mass / (rsq * sqrt(rsq)));
-        target.applyAcceleration(i, accel);
+    // Loop unrolling with a factor of 8
+    const int UNROLL_FACTOR = 8;
+    const Vector3D<Real>& centroid = source.data.centroid;
+    const Real& sum_mass = source.data.sum_mass;
+    Vector3D<Real> diff[UNROLL_FACTOR];
+    Real rsq[UNROLL_FACTOR];
+    Vector3D<Real> accel[UNROLL_FACTOR];
+
+    for (int i = 0; i < target.n_particles; i += UNROLL_FACTOR) {
+      diff[0] = centroid - target.particles()[i].position;
+      diff[1] = centroid - target.particles()[i+1].position;
+      diff[2] = centroid - target.particles()[i+2].position;
+      diff[3] = centroid - target.particles()[i+3].position;
+      diff[4] = centroid - target.particles()[i+4].position;
+      diff[5] = centroid - target.particles()[i+5].position;
+      diff[6] = centroid - target.particles()[i+6].position;
+      diff[7] = centroid - target.particles()[i+7].position;
+      rsq[0] = diff[0].lengthSquared();
+      rsq[1] = diff[1].lengthSquared();
+      rsq[2] = diff[2].lengthSquared();
+      rsq[3] = diff[3].lengthSquared();
+      rsq[4] = diff[4].lengthSquared();
+      rsq[5] = diff[5].lengthSquared();
+      rsq[6] = diff[6].lengthSquared();
+      rsq[7] = diff[7].lengthSquared();
+      if (rsq[0] != 0) {
```

Comment:
Not so sure about the efficiency of these if statements though.

Reply:
we might be able to filter out rsq=0s earlier? also we can probably assume that rsq=0 is only true for one particle since the particles are in different locations.